File: client-spring-boot/src/main/java/grgic/antonio/client_spring_boot/service/RestProxyService.java
package grgic.antonio.client_spring_boot.service;

import grgic.antonio.client_spring_boot.model.LargeObject;
import grgic.antonio.client_spring_boot.model.MediumObject;
import grgic.antonio.client_spring_boot.model.SmallObject;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestClient;

@Service
public class RestProxyService {

    private final RestClient javaClient;

    public RestProxyService(
            @Value("${client.rest-java.base-url}") String javaBaseUrl) {
        this.javaClient = RestClient.create(javaBaseUrl);
    }

    public void fetch(String size) {
        switch (size) {
            case "small"  -> javaClient.get().uri("/api/payload/small").retrieve().body(SmallObject.class);
            case "medium" -> javaClient.get().uri("/api/payload/medium").retrieve().body(MediumObject.class);
            case "large"  -> javaClient.get().uri("/api/payload/large").retrieve().body(LargeObject.class);
            default -> throw new IllegalArgumentException("Unknown size: " + size);
        }
    }
}

File: client-node/src/service/RestProxyService.js
const { SmallObject }  = require('../model/SmallObject');
const { MediumObject } = require('../model/MediumObject');
const { LargeObject }  = require('../model/LargeObject');

const BASE_URL = process.env.REST_NODE_URL || 'http://localhost:3000';

async function fetchFromServer(size) {
    const url = `${BASE_URL}/api/payload/${size}`;
    const res = await fetch(url);

    if (!res.ok) {
        const err = new Error(`REST upstream returned ${res.status} ${res.statusText} for ${url}`);
        err.status = res.status;
        throw err;
    }

    const data = await res.json();

    switch (size) {
        case 'small':  return new SmallObject(data);
        case 'medium': return new MediumObject(data);
        case 'large':  return new LargeObject(data);
        default: throw new Error(`Unknown size: ${size}`);
    }
}

module.exports = { fetchFromServer };
